## 서블릿 예외 처리 - 필터

목표

예외 처리에 따른 필터와 인터셉터 그리고 서블릿이 제공하는 `DispatchType` 이해하기

예외 발생과 오류 페이지 요청 흐름

```
1. WAS(여기까지 전파) <- 필터 <- 서블릿 <- 인터셉터 <- 컨트롤러(예외발생)

2. WAS `/error-page/500` 다시 요청 -> 필터 -> 서블릿 -> 인터셉터 -> 
컨트롤러(/errorpage/500) -> View
```

<br/>

### 오류가 발생하면 오류 페이지를 출력하기 위해 WAS 내부에서 다시 한번 호출이 발생한다. <br/>이때 필터, 서블릿, 인터셉터도 모두 다시 호출된다.

<br/>

그런데 로그인 인증 체크 같은 경우를 생각해보면, 이미 한번 필터나, 인터셉터에서 로그인 체크를 완료했다. 

### 따라서 서버 내부에서 오류 페이지를 호출한다고 해서 해당 필터나 인터셉트가 다시 한번 더 <br/>호출되는 것은 매우 비효율적이다.

<br/>

결국 클라이언트로 부터 발생한 정상 요청인지, 아니면 오류 페이지를 출력하기 위한 내부 요청인지 구분할 수 있어야 한다. 

서블릿은 이런 문제를 해결하기 위해 DispatcherType 이라는 추가 정보를 제공한다.

<br/>

이렇게 비효율적인 부분을 제거 하였다.

![이미지](/programming/img/나32.PNG)

<br/>

## DispatcherType

필터는 이런 경우를 위해서 `dispatcherTypes` 라는 옵션을 제공한다.

이전 강의의 마지막에 다음 로그를 추가했다.

<br/>

`log.info("dispatchType={}", request.getDispatcherType())`

그리고 출력해보면 오류 페이지에서 `dispatchType=ERROR` 로 나오는 것을 확인할 수 있다.

고객이 처음 요청하면 `dispatcherType=REQUEST` 이다.

<br/>

이렇듯 서블릿 스펙은 실제 고객이 요청한 것인지, 서버가 내부에서 오류 페이지를 요청하는

것인지 `DispatcherType` 으로 구분할 수 있는 방법을 제공한다.

<br/>

`javax.servlet.DispatcherType`

```java
public enum DispatcherType {
 FORWARD,
 INCLUDE,
 REQUEST,
 ASYNC,
 ERROR
}
```

### DispatcherType

- `REQUEST` : 클라이언트 요청

- `ERROR` : 오류 요청
- `FORWARD` : MVC에서 배웠던 서블릿에서 다른 서블릿이나 JSP를 호출할 때
    - RequestDispatcher.forward(request, response);
- `INCLUDE` : 서블릿에서 다른 서블릿이나 JSP의 결과를 포함할 때
    - RequestDispatcher.include(request, response);
- `ASYNC` : 서블릿 비동기 호출

<br/>

## 흐름 설명

![이미지](/programming/img/나33.PNG)

1. 첫번째 줄 코드 `LogFilter` 클래스
2. 그 다음 두번째 줄에서 `chain.doFilter` 했더니 

<br/>

```java
try {
      log.info("REQUEST [{}][{}][{}]", uuid, request.getDispatcherType(), requestURI);
      chain.doFilter(request, response);
```


3. `ServletExController` 컨트롤러에서 `errorEx()` 메서드 실행하여 ‘예외 발생’
    
    이유는? request 에 클라이언트의 요청 URL이 들어 있기 때문.
    
4. 그리고 예외가 터졌더니, 다시 서블릿으로 올라온다. 

<br/>

```java
public class ServletExController {

    @GetMapping("/error-ex")
    public void errorEx() {
        throw new RuntimeException("예외 발생");
		}

		``
}
```

5. 그러면 `catch` 부분에서 컨트롤러에서 예외 터진 걸 잡을 것이다.

6. 그렇게 한것이 콘솔 2번째 줄인 것이다.

<br/>

```java
} catch (Exception e) {
     log.info("EXCEPTION {}", e.getMessage());
     throw e;
}
```


7. 그 다음 finally 가 무조건 찍히는 것이다.

8. 그게 콘솔창 3번째 줄 인 것이다.

<br/>

```java
} finally {
      log.info("RESPONSE [{}][{}][{}]", uuid, request.getDispatcherType(), requestURI);
  }
```

<br/>

9. 그 다음, 일단 `throw e;` 라고 던졌으니, WAS까지 올라가는 것이다.



10. 그러면 WAS에서 콘솔창에 이렇게 출력한다.


![이미지](/programming/img/나34.PNG)

<br/>

11. 그렇게 콘솔창에 찍은 다음에 다시 WAS에서 내려 오는 것이다. (밑에 사진 참고)


![이미지](/programming/img/나35.PNG)

<br/>

12. 그렇게 다시 밑에 코드 로직을 타게 되는 것이다.


```java
try {
      log.info("REQUEST [{}][{}][{}]", uuid, request.getDispatcherType(), requestURI);
      chain.doFilter(request, response);

```

13. 그런데 WAS에서 내려온 다음에  [/error-page/500] 이 경로로 요청을 한다. 왜??
14. 밑에 내가 만들었던 `WebServerCustomizer` 클래스에서 `RuntimeException` 에러가 터졌다면 
    
    `"/error-page/500"` 으로 가라! 했기 때문이다.
    
    ### 하지만 여기서 중요한 부분은 (11번) 밑에 그림을 보면 디스패처 타입이 
    [ERROR] 로 변경 되었다는 것을 알아야 된다.

<br/>
    

```java
public class WebServerCustomizer implements ``

@Override
public void customize(ConfigurableWebServerFactory factory) {

						``

      ErrorPage errorPageEx = new ErrorPage(RuntimeException.class, "/error-page/500");

      factory.addErrorPages(errorPage404, errorPage500, errorPageEx);
 }
```

15. 그 다음 다시 `chain.doFilter(request**,**response)`를 호출하게 되는 것이다.

16. 밑에 코드 처럼 `ErrorPageController` 컨트롤러 errorPage500() 메서드를 호출하게 된다.
17. 그렇게 하여 (11번) 밑 사진에 두번째 줄 `"errorPage 500"` 가 출력 되는 것이다.
18. 그리고 밑에 `printErrorInfo(request);` 메서드를 실행 시키면서.
19. `printErrorInfo` 메서드로 이동하게 된다.

```java
public class ErrorPageController {
	
	``

@RequestMapping("/error-page/500")
public String errorPage500(HttpServletRequest request,
							HttpServletResponse response) {

		log.info("errorPage 500");
		
		printErrorInfo(request);
				return "error-page/500";
}
```

<br/>

20. 이동하여 이렇게 모두 출력 시키는 것이다. 밑에 사진으로 더 자세히 보기.

```java
private void printErrorInfo(HttpServletRequest request) {
        log.info("ERROR_EXCEPTION: ex=", request.getAttribute(ERROR_EXCEPTION));
        log.info("ERROR_EXCEPTION_TYPE: {}", request.getAttribute(ERROR_EXCEPTION_TYPE));
        log.info("ERROR_MESSAGE: {}", request.getAttribute(ERROR_MESSAGE)); //ex의 경우 NestedServletException 스프링이 한번 감싸서 반환
        log.info("ERROR_REQUEST_URI: {}", request.getAttribute(ERROR_REQUEST_URI));
        log.info("ERROR_SERVLET_NAME: {}", request.getAttribute(ERROR_SERVLET_NAME));
        log.info("ERROR_STATUS_CODE: {}", request.getAttribute(ERROR_STATUS_CODE));
        log.info("dispatchType={}", request.getDispatcherType());
    }
```

<br/>

![이미지](/programming/img/나36.PNG)

21. 그리고 다시 돌아와 `return "error-page/500";` html 코드로 가서 
    웹 브라우저에 뿌리고 끝나는 것이다.
    
<br/>


전체 화면

![이미지](/programming/img/나37.PNG)

<br/>

## 만약 밑에 코드에서 DispatcherType.ERROR 뺀다면??

```java
@Bean 
public FilterRegistrationBean logFilter() {

		```

filterRegistrationBean.setDispatcherTypes(DispatcherType.REQUEST);

		```
}
```
<br/>

여기서 만약 DispatcherType.ERROR 를 뺀다면 바로 

`chain.doFilter(request, response);` 로 가게 되는 것이다. 

<br/>

무슨말인가?

여기 로직에서 `log.info("REQUEST ``` )` 부분을 타지 않고 바로 doFilter 로 간다는 뜻이다.


```java
try {
      log.info("REQUEST [{}][{}][{}]", uuid, request.getDispatcherType(), requestURI);
      chain.doFilter(request, response);
```


순서를 보자면 이렇다.

두번째 호출은 사실 정상 호출은 아니다. 

오류 상황이고, “이 오류상황을 애플리케이션에서 어떻게 보여줄꺼야?” 라고 물어보는 것이다.

```java
WAS /error-page/500 다시 요청 -> 필터 -> 서블릿 -> 
인터셉터 -> 컨트롤러(/error-page/500) -> View
```

<br/>

[ERROR] 가 없는 것을 알 수 있다.

![이미지](/programming/img/나38.PNG)

<br/>

## 참고 사항

이거 같은 경우엔 `REQUEST`, `ERROR` 두 가지 경우에 

`FilterRegistrationBean` 가 호출이 되는 것이다. 

즉, 이거 같은 경우에 실행 시켜보면 logFilter() 메서드가 두번 호출 되는 것을 알 수 있다.

```java
@Bean 
public FilterRegistrationBean logFilter() {

		``

filterRegistrationBean.setDispatcherTypes(DispatcherType.REQUEST,
			                                DispatcherType.ERROR);
		``

}
```

<br/>

>**Reference** <br/>스프링 MVC 2편 - 백엔드 웹 개발 활용 기술 - https://www.inflearn.com/course/%EC%8A%A4%ED%94%84%EB%A7%81-mvc-2